runtime/bcutil/defineclass.c

```diff
@@ -861,4 +885,111 @@ classCouldPossiblyBeShared(J9VMThread * vmThread, J9LoadROMClassData * loadData)
+		const U_8 *hostClassNameData = J9UTF8_DATA(hostClassName);
+		const U_8 *anonClassNameData = J9UTF8_DATA(anonClassName);
+		const UDATA hostClassNameLength = J9UTF8_LENGTH(hostClassName);
+
+		/* Anonymous class name has trailing digits. Example - "test/DummyClass/00000000442F098".
+		 * The code below removes the trailing digits, "/00000000442F098", from the anonymous class name.
+		 */
+		UDATA anonClassNameLength = J9UTF8_LENGTH(anonClassName) - 1;
+		for (; anonClassNameLength >= 0; anonClassNameLength--) {
+			if (anonClassNameData[anonClassNameLength] == '/') {
+				break;
+			}
+		}
+
+		bufLen = j9str_printf(PORTLIB, NULL, 0, errorMsg,
+						hostClassNameLength, hostClassNameData,
+						anonClassNameLength, anonClassNameData);
+		if (bufLen > 0) {
+			buf = j9mem_allocate_memory(bufLen, OMRMEM_CATEGORY_VM);
+			if (NULL != buf) {
+				j9str_printf(PORTLIB, buf, bufLen, errorMsg,
+						hostClassNameLength, hostClassNameData,
+						anonClassNameLength, anonClassNameData);
+			}
+		}
+	}
+
+	return buf;
+}
+
+/* From Java 9 and onwards, set IllegalArgumentException when host class and anonymous class have different packages. */
+static void
+setIllegalArgumentExceptionHostClassAnonClassHaveDifferentPackages(J9VMThread *vmStruct, J9ROMClass *anonROMClass, J9ROMClass *hostROMClass) {
+	PORT_ACCESS_FROM_VMC(vmStruct);
+	const J9JavaVM *vm = vmStruct->javaVM;
+	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+
+	/* Construct error string */
+	const char *errorMsg = j9nls_lookup_message(J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE, J9NLS_JCL_HOSTCLASS_ANONCLASS_DIFFERENT_PACKAGES, NULL);
+	char *buf = createErrorMessage(vmStruct, anonROMClass, hostROMClass, errorMsg);
+	vmFuncs->setCurrentExceptionUTF(vmStruct, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, buf);
+	j9mem_free_memory(buf);
+}
+
+/* Free the memory segment corresponding to the anonymous ROM class. */
+static void
+freeAnonROMClass(J9JavaVM *vm, J9ROMClass *romClass) {
+	if (NULL != romClass) {
+		J9MemorySegment **previousSegmentPointerROM = &vm->anonClassLoader->classSegments;
+		J9MemorySegment *segmentROM = *previousSegmentPointerROM;
+
+		/* Walk all anonymous classloader's ROM memory segments. If ROM class
+		 * is allocated there it would be one per segment.
+		 */
+		while (NULL != segmentROM) {
+			J9MemorySegment *nextSegmentROM = segmentROM->nextSegmentInClassLoader;
+			if (J9_ARE_ALL_BITS_SET(segmentROM->type, MEMORY_TYPE_ROM_CLASS)
+				&& ((J9ROMClass *)segmentROM->heapBase == romClass)
+			) {
+				/* Found memory segment corresponding to the ROM class. Remove
+				 * this memory segment from the list.
+				 */
+				*previousSegmentPointerROM = nextSegmentROM;
+				/* Free memory segment corresponding to the ROM class. */
+				vm->internalVMFunctions->freeMemorySegment(vm, segmentROM, 1);
+				break;
+			}
+			previousSegmentPointerROM = &segmentROM->nextSegmentInClassLoader;
+			segmentROM = nextSegmentROM;
+		}
```

Comment:
Please add an Trc_BCU_Assert_ShouldNeverHappen or similar if the memory segment is not found.

Reply:
added.